### PR TITLE
Add AbstractThrowableAssert#causesChain to simplify performing assertions on a chain of throwables

### LINKED ISSUE #1147: Provide semantic ways to test n-level (n>1) causes of an Exception, not only direct cause and root cause
Assertj currently provides methods to test a thrown exception, its direct cause, or its root cause. It does not offer a semantic way to test the exceptions/causes in between the first cause and the root cause.

Exceptions tend to get wrapped and wrapped by frameworks and libraries, and by project code. Sometimes the most meaningful exception is neither the one thrown nor its cause nor the root cause, but one somewhere in the middle.

For example, BusinessLogicEx caused by SqlRollbackEx caused by UnableToExecuteSqlEx in a test where I just want to assert that my sql will be rolled back by finding the SqlRollbackEx in the stack.

```java
assertThatThrownBy(() -> {}).hasAnyLevelCauseInstanceOf(type);
assertThatThrownBy(() -> {}).hasAnyLevelCauseWithMessageMatching(condition);
// etc
```

### PR DESCRIPTION
**Pull request by slinkydeveloper**
#### Check List:
* Fixes #1147 (ignore if not applicable)
* Unit tests : YES
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md)

Following the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md) will make it easier for us to review and accept your PR.

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
Hi @joel-costigliola I stumbled on the issue #1147 and I created this PR to propose a solution, lmk what do you think. I'm a first time contributor here, so I hope I haven't messed it up badly :smile:

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java [hunk 1]
```diff
@@ -203,6 +203,31 @@ public SELF hasNoCause() {
     return new ThrowableAssert<>(org.assertj.core.util.Throwables.getRootCause(actual));
   }
 
+  /**
+   * Returns a new list assertion of every cause in the chain of causes.
+   * The list is ordered from the current {@link Throwable} up to the root cause.
```
Comment (matriv) [inline src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:208]: Myself, I kinda prefer the term `down to the root cause`.

File: src/main/java/org/assertj/core/util/Throwables.java [hunk 2]
```diff
@@ -148,6 +150,21 @@ public static Throwable getRootCause(Throwable throwable) {
     return throwable;
   }
 
+  /**
+   * Get the causes chain of a {@link Throwable}.
+   *
+   * @param throwable the {@code Throwable} to get causes chain from.
+   * @return the causes chain, including the {@link Throwable} provided as parameter.
+   */
+  public static List<Throwable> getCausesChain(Throwable throwable) {
```
Comment (matriv) [inline src/main/java/org/assertj/core/util/Throwables.java:159]: I'm kind of unsettled if the `throwable` in test should be included or not in the list.
On one hand it's nice to have the complete list and assert any of the throwables in the chain, on the other hand, one might say that `getCausesChain` only contains causes down from the "current" level, excluding the current level. Moreover, it's easier to assert the `throwable` in question separately rather than excluding it from the list manually if the user doesn't want to assert it.

I have a slight preference to keep it as is (including the `throwable` in test), but I'd say to specifically mention the inclusive behaviour in the javadocs, here and in `asCausesChain()` method.

**Review by matriv — COMMENTED** [review]
Thank you @slinkydeveloper! That's a very useful functionality for asserting exceptions!

File: src/main/java/org/assertj/core/util/Throwables.java (hunk 2 above)
Comment (scordio) [inline src/main/java/org/assertj/core/util/Throwables.java:159]: I don't see the need of exposing this as a public API, can be a private method of `AbstractThrowableAssert`.

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java [hunk 3]
```diff
@@ -203,6 +203,31 @@ public SELF hasNoCause() {
...
+   * Throwable throwable = new Exception("boom!", cause);
+   *
+   * // typical use:
+   * assertThat(throwable).asCausesChain()
+   *                      .extracting(Throwable::getMessage)
+   *                      .anyMatch("some failure");</code></pre>
+   *
+   * @return a new assertion object
+   * @throws AssertionError if the actual {@code Throwable} is {@code null}.
+   * @throws AssertionError if the actual {@code Throwable} does not have a cause.
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:222]: I'm not sure if we should relax this and return an empty list instead.

File: src/main/java/org/assertj/core/util/Throwables.java (hunk 2 above)
Comment (scordio) [reply src/main/java/org/assertj/core/util/Throwables.java:159]: I tend to favor not having `actual` in the resulting list, starting with its direct cause instead.

**Review by scordio — COMMENTED** [review]
Thanks @slinkydeveloper, some initial thoughts from my side. We would need some additional work to make sure everything works fine with soft assertions, but let's have first feedback from @joel-costigliola about naming and general direction.

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java (hunk 3 above)
Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:222]: It's not an obvious call  but I think I'm in favor of returning an empty list.

pros: 
- you can write `assertThat(throwable).causesChain().isEmpty();` which is reasonable
- in API I tend to prefer returning empty collection over null/failure

cons: 
- if we return an empty list then some assertions like `allMatch` will pass (it succeeds for any empty list)
- users probably expect to have causes some failing if there are none might be better

The last point is not a real issue IMO as the assertions applied on the list of causes will most probably fail if we return an empty list (except for a few like `allMatch`).

TLDR: let's go with returning an empty list ;-)

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java (hunk 1 above)
Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:208]: `Returns a new list assertion of every cause down to the root cause. The list is empty if there is no cause.`  sounds good to me

File: src/main/java/org/assertj/core/util/Throwables.java (hunk 2 above)
Comment (joel-costigliola) [reply src/main/java/org/assertj/core/util/Throwables.java:159]: When I first read `causeChain` I thought it would only contain causes and not the initial exception so I would leave it out.
We can always add a `causeChainWithInitialThrowable` later on if users ask for it.

Comment (joel-costigliola) [reply src/main/java/org/assertj/core/util/Throwables.java:159]: Agreed, we'll move it here if we reuse that piece of code elsewhere.

File: src/test/java/org/assertj/core/api/throwable/ThrowableAssert_asCausesChain_Test.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  private final Throwable rootCause = new NullPointerException("root");
+  private final Throwable intermediateCause = new IllegalStateException("intermediate", rootCause);
+
+  @Test
+  void should_return_throwable_chain_assertion() {
+    // GIVEN
+    Throwable throwable = new Throwable("initial", intermediateCause);
+    // WHEN/THEN
+    assertThat(throwable).asCausesChain()
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/throwable/ThrowableAssert_asCausesChain_Test.java:32]: minor: we prefer `then` here (from `BDDAssertions.then` static import)

**Comment by joel-costigliola** [comment]
@slinkydeveloper we have reviewed  your PR, thanks for the good work!

File: src/main/java/org/assertj/core/util/Throwables.java (hunk 2 above)
Comment (slinkydeveloper) [reply src/main/java/org/assertj/core/util/Throwables.java:159]: May I propose to introduce both in the same PR? Perhaps one could be `causesChain` and the other one `throwableChain` or `throwableTree`, which includes the initial one.

File: src/test/java/org/assertj/core/api/throwable/ThrowableAssert_asCausesChain_Test.java (hunk 4 above)
Comment (slinkydeveloper) [reply src/test/java/org/assertj/core/api/throwable/ThrowableAssert_asCausesChain_Test.java:32]: WDYM? I honestly don't get how you want me to refactor this particular line of code :smile: I modified the code in the tests below to use `BDDAssertions.then`

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java (hunk 3 above)
Comment (slinkydeveloper) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:222]: Modified with the empty list

File: src/main/java/org/assertj/core/api/AbstractThrowableAssert.java [hunk 5]
```diff
@@ -203,6 +203,31 @@ public SELF hasNoCause() {
...
+   *
+   * @since 3.16.0
+   */
+  public ListAssert<Throwable> asCausesChain() {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply src/main/java/org/assertj/core/api/AbstractThrowableAssert.java:226]: How would you suggest to change the return type? I changed it to `AbstractListAssert<?, List<Throwable>, Throwable, ? extends AbstractThrowableAssert<?, Throwable>>`, although i'm not completely sure it's the correct one

**Comment by joel-costigliola** [comment]
This fell under our radar, let's schedule it for the 4.0 release.

**Comment by joel-costigliola** [comment]
Integrated thanks @slinkydeveloper, I changed a bit the naming to `throwablesChain` to make it clear it's not only the causes but also the root exception.
